File: addons/imgui-godot/ImGuiGodot/Internal/InputLocal.cs
#if GODOT_PC
using Godot;
using ImGuiNET;

namespace ImGuiGodot.Internal;

internal sealed class InputLocal : Input
{
    protected override void UpdateMousePos(ImGuiIOPtr io)
    {
        // do not use global mouse position
    }

    public override bool ProcessInput(InputEvent evt)
    {
        // no support for SubViewport widgets

        if (evt is InputEventMouseMotion mm)
        {
            var io = ImGui.GetIO();
            var mousePos = mm.Position;
            io.AddMousePosEvent(mousePos.X, mousePos.Y);
            mm.Dispose();
            return io.WantCaptureMouse;
        }
        return HandleEvent(evt);
    }
}
#endif

File: gdext/src/InputLocal.cpp
#include "InputLocal.h"
#include <godot_cpp/classes/input_event_mouse_motion.hpp>
#include <imgui.h>

namespace ImGui::Godot {

void InputLocal::UpdateMousePos()
{
    // do not use global mouse position
}

bool InputLocal::ProcessInput(const Ref<InputEvent>& evt)
{
    // no support for SubViewport widgets

    if (Ref<InputEventMouseMotion> mm = evt; mm.is_valid())
    {
        ImGuiIO& io = ImGui::GetIO();
        Vector2 mousePos = mm->get_position();
        io.AddMousePosEvent(mousePos.x, mousePos.y);
        return io.WantCaptureMouse;
    }
    return HandleEvent(evt);
}

} // namespace ImGui::Godot
